lib/src/markdown_content.dart

```diff
@@ -9,67 +9,56 @@ import 'package:markdown/markdown.dart';
 import 'package:meta/meta.dart';
 import 'package:html/parser.dart' as html_parser;
 import 'package:path/path.dart' as p;
+import 'package:source_span/source_span.dart';
 
 import 'model.dart';
 
 /// The extracted content of a markdown file.
-class ScannedMarkdownContent {
-  String parseError;
-  List<String> images;
-  List<String> links;
+class ExctractedMarkdownContent {
+  final List<Link> images;
+  final List<Link> links;
+
+  ExctractedMarkdownContent({this.images, this.links});
 
   Map<String, dynamic> toJson() => <String, dynamic>{
-        'parseError': parseError,
-        'images': images,
-        'links': links,
+        'images': images.map((l) => l.url).toList(),
+        'links': links.map((l) => l.url).toList(),
       };
 }
 
 /// Scans a markdown text and extracts its content.
-ScannedMarkdownContent scanMarkdownText(String text) {
-  try {
-    final htmlText = markdownToHtml(text);
-    final html = html_parser.parseFragment(htmlText);
-    final images = html
-        .querySelectorAll('img')
-        .where((e) => e.attributes.containsKey('src'))
-        .map((e) => e.attributes['src'])
-        .toList();
-    final links = html
-        .querySelectorAll('a')
-        .where((e) => e.attributes.containsKey('href'))
-        .map((e) => e.attributes['href'])
-        .toList();
-    return ScannedMarkdownContent()
-      ..images = Set<String>.from(images).toList()
-      ..links = Set<String>.from(links).toList();
-  } catch (e) {
-    return ScannedMarkdownContent()..parseError = e.toString();
-  }
+ExctractedMarkdownContent scanMarkdownText(String text, Uri sourceUrl) {
+  final htmlText = markdownToHtml(text);
+  final html = html_parser.parseFragment(htmlText,
+      sourceUrl: sourceUrl.toString(), generateSpans: true);
+  return ExctractedMarkdownContent(
+      images: _unique(html
+          .querySelectorAll('img')
+          .where((e) => e.attributes.containsKey('src'))
+          .map((e) => Link(e.attributes['src'], e.sourceSpan))),
+      links: _unique(html
+          .querySelectorAll('a')
+          .where((e) => e.attributes.containsKey('href'))
+          .map((e) => Link(e.attributes['href'], e.sourceSpan))));
 }
 
+List<T> _unique<T>(Iterable<T> l) => [
```

Comment:
nit: `=> l.toSet().toList()` would be easier to parse for me

Reply:
Done